### ISSUE TITLE
spdk_thread_send_msg error handling

### ISSUE THREAD

**Issue by mhae** [spdk/spdk#811]
It seems that spdk_thread_send_msg can fail if there are no more entries in the ring buffer holding the messages (see https://github.com/spdk/spdk/blob/264667078184117dfc265fa525276356d0f69f9a/lib/thread/thread.c#L696)

While the ring buffer is fairly large 64k, there can still be situations where a thread is sending messages to another thread at a higher rate than it can handle.
If I understand the DPDK rte_ring functionality correctly, then it is not blocking and we don't really have any prevention.

## Expected Behavior

Some form of letting the caller know that a message couldn't be queued.

**Comment by benlwalker** [comment]
Correct - we need to figure out what our strategy is going to be here. Either we need to let the message pass fail and then try to handle failures throughout the code, or we need to possibly block the thread and spin until it works. Both aren't great options, but fortunately there isn't that much message passing going on in SPDK so it should be tractable to resolve.

**Comment by mhae** [comment]
Agreed, for the basic SPDK use cases it is probably okay ... but issues can arise when we use SPDK threading in a different execution env.
What do you think about adding a function that returns whether the message queue is full? We already have spdk_thread_is_idle that looks at the ring count.
This at least would allow a caller to avoid posting more messages when the queue is full.

**Comment by ak-gh** [comment]
On Thu, Jun 6, 2019, 19:35 Ben Walker <notifications@github.com> wrote:

> Correct - we need to figure out what our strategy is going to be here.
> Either we need to let the message pass fail and then try to handle failures
> throughout the code, or we need to possibly block the thread and spin until
> it works.
>
There's a third option thanks to us owning the message data structure. It
should be relatively straightforward to add a tailq link field to the
message/tailq head to the thread structure so as, on ring_enqueue failure,
to queue message in a per-thread tailq. Queued messages could be retried
when, say, the thread receives a message (which may be considered as an
indication that ring space pressure has subsided).

Regards,
Andrey

Both aren't great options, but fortunately there isn't that much message

**Comment by benlwalker** [comment]
That's an interesting idea and I like it. I need to think about exactly when to retry and if there are any pathological cases.

Simultaneously adding a function to query if the ring is full would be great too.

**Comment by ak-gh** [comment]
On Fri, Jun 7, 2019, 22:16 Ben Walker <notifications@github.com> wrote:

> That's an interesting idea and I like it. I need to think about exactly
> when to retry and if there are any pathological cases.
>
Most natural retry point is when a message from the thread the receiver has
queueed messages for is received, though that would require a search tree
keyed by the destination tid instead of the plain tailq. My gut feeling,
though, is that in a reasonably well-connected topology receiving any
message should be good enough to retry.

> Simultaneously adding a function to query if the ring is full would be
> great too.
>
Given that dpdk rings are non-blocking, I'm not sure if there's any
straightforward way to check that reliably.

Regards,
Andrey

**Comment by mhae** [comment]
Either is fine... I still think it is good to have an API to check the # of messages queued even if it is not 100% reliable. It allows me to see what's going on with submissions and expose a metric.

**Comment by jimharris** [comment]
Hi @mhae - we chatted briefly about this issue at our last bug scrub.

A first patch could just be changing spdk_thread_send_msg() to return int instead of void.  Then at least the caller knows if the message ring was full and the message couldn't be sent.

Next we can either move the #defines, or provide getter functions, for the default message ring size.  I'd lean toward the getter function, just in case we make this run-time configurable in the future.

Finally we can add a function that gets the current count of messages in the ring (using spdk_ring_count).

I'd prefer to not add any kind of queueing and retry logic until we have a strong use case for it in the upstream SPDK tree.

What do you think?

**Comment by mhae** [comment]
@jimharris , returning an int is totally find. This will allow the caller to deal with this condition. I don't think we need a retry logic in spdk_thread_send_msg() either. Just let the caller know that there is an issue suffices for now.

**Comment by jimharris** [comment]
https://review.gerrithub.io/c/spdk/spdk/+/472438

**Comment by mhae** [comment]
@jimharris Awesome! Looks good. Thanks!

**Comment by jimharris** [comment]
Patch merged.  Thanks @mhae for the report!

**Comment by skypexu** [comment]
This really does not fix the problem. Because  SPDK_MSG_MEMPOOL_CACHE_SIZE is 1024, this means per thread can have at most 1024 entries cached, and g_spdk_msg_mempool only has 65536 entries, so 10 threads caches 10240 entries, 65 threads can exhaust the messages resources!

**Comment by jimharris** [comment]
> This really does not fix the problem. Because SPDK_MSG_MEMPOOL_CACHE_SIZE is 1024, this means per thread can have at most 1024 entries cached, and g_spdk_msg_mempool only has 65536 entries, so 10 threads caches 10240 entries, 65 threads can exhaust the messages resources!

There are ways to configure the number of messages allocated in the pool (spdk_thread_lib_init_ext(), spdk_app_opts::msg_mempool_size), which partially solves this problem if the developer is using those APIs already.  The thread library does not export anything about the number of msg objects it caches per spdk thread, which the developer would need to understand how to size the mempool.

For in-tree applications like spdk_tgt and nvmf_tgt, there is no logic in place currently to autosize these defaults up if there are too many cores defined in the core mask.  Nor is there a command line option to specify the number of message objects.  Personally I think the former would be preferred, the SPDK app framework knows the core mask and can determine the right msg pool size rather than requiring the user to figure it out and specify it on the command line.

I'm going to reopen this and mark it as an enhancement request for now.

**Comment by m-spiewak** [comment]
@jimharris What would be the right, automatically calculated size of g_spdk_msg_mempool?
Is 'core_count * 1024'  enough ? Or multiply this by some factor (e.g. 4 or 8) to have some spare room?
Currently it is created with fixed size of SPDK_DEFAULT_MSG_MEMPOOL_SIZE, this means 262143

**Comment by jimharris** [comment]
> @jimharris What would be the right, automatically calculated size of g_spdk_msg_mempool?
> Is 'core_count * 1024' enough ? Or multiply this by some factor (e.g. 4 or 8) to have some spare room?
> Currently it is created with fixed size of SPDK_DEFAULT_MSG_MEMPOOL_SIZE, this means 262143

I think factor of 4 is needed.  We want some spare room, we also need to account for potentially multiple spdk_threads per core - for example, iscsi + nvmf + vhost all running in spdk_tgt.

I'd suggest keeping existing MSG_MEMPOOL_SIZE as the default when num cores is less than 64, but when core count gets bigger, we increase msg mempool size accordingly.

As a follow-on patch (or patches), we should improve thread.c's usage of the spdk_mempool.  For example, when cache is exceeded, we will always just spdk_mempool_put one message at a time.  Similarly on get when the cache is empty.  But it is actually better to be somewhat relaxed on the cache size.  For example, let the cache grow slightly bigger than the max size, if it gets 8 or 16 above the max, then put 8 or 16 back into the mempool at once.  Similarly on get, if the cache is empty, don't just get 1 message a time, get 8 or 16, use 1, put the rest into the cache.  This will avoid unnecessary contention on the lock guarding the mempool.  Could you look at this too @m-spiewak?

**Comment by m-spiewak** [comment]
@jimharris Sure, I will look at that.
